### PR TITLE
cups-ppdc: new port

### PR DESCRIPTION
**Pull request by evanmiller**
#### Description

`cups-ppdc` includes a collection of command-line utilities for creating and processing PostScript Printer Description files. These utilities are found in CUPS on OS X 10.6 and later in `/usr/bin`, and are being packaged here for the benefit of users of earlier OSes. Note that these tools have been deprecated by Apple as of 2018, and are likely to be removed from a future version of macOS. They are essentially text processing tools and can be used without support from the operating system.

The included tools are:

* `ppdc` - Compile a driver (.drv) file to multiple PPDs
* `ppdi` - Import one or more PPD files into a driver file
* `ppdhtml` - Summarize a driver file as an HTML page
* `ppdpo` - Extract localizable messages from a driver file
* `ppdmerge` - Merge multiple PPD files into one

These utilities are commonly grouped together (see e.g. [`cups-ppdc`](https://packages.debian.org/sid/cups-ppdc) from Debian) as a subset of CUPS. Unlike the Debian package, this package does not install or require a dynamic library, as the binaries here are all statically linked.

Man pages are included in the installation. My personal use-case was getting a modern USB printer working on an old iMac.

###### Type(s)

- [ ] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on

macOS 10.4.11 8S165 Power Macintosh
Component versions: DevToolsCore-798.0; DevToolsSupport-794.0

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? 
- [x] checked your Portfile with `port lint`?
- [ ] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

### CODE REVIEW

**Comment by evanmiller** [comment]
Quick update. The original submission included the tools from CUPS 1.6 (corresponding to OS X 10.8). I was able to get the latest CUPS release from Apple to compile, so this package now makes PPD tools from CUPS 2.3.3 available on all OS versions.

Currently the tools are statically compiled, but a logical next step would be to make a `libcups2` package that includes the CUPS dynamic library and headers.

File: print/cups-ppdc/Portfile [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+license         Apache-2
+maintainers     nomaintainer
+description     Compiler for PostScript Printer Description (PPD) files
+long_description \
+    cups-ppdc is a collection of utilities taken from the Common UNIX Printing System (CUPS) \
+    for working with PostScript Printer Description (PPD) files\; \
+    PPD files can be used to set up a printer when proprietary drivers are not available. \
+    See the ppdcfile(5) man page for information on the PPD file format.
+
+homepage        https://github.com/apple/cups
```
Comment (reneeotten) [inline print/cups-ppdc/Portfile:19]: this is the default set by the github PG, so the line is redundant

File: print/cups-ppdc/Portfile [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    if {${os.platform} eq "darwin" && ${os.major} < 16} {
+        reinplace "s|<os/availability.h>|<AvailabilityMacros.h>|" cups/versioning.h
+    }
+}
+
+configure.args-append   --disable-dnssd
+configure.args-append   --disable-gssapi
+configure.args-append   --disable-ssl
+configure.args-append   --disable-shared
+configure.args-append   --with-bundledir=""
```
Comment (reneeotten) [inline print/cups-ppdc/Portfile:51]: please use a line continuation instead of repeating `configure.args-append`

File: print/cups-ppdc/Portfile [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    platform darwin 8 {
+        # res_init() symbol is found, but the prototype is never included due
+        # to broken resolv.h on Tiger. Don't need networking so just tell the
+        # build we've never heard of res_init().
+        reinplace "s|#define HAVE_RES_INIT 1|/* #undef HAVE_RES_INIT */|" config.h
+    }
+}
+
+build {
+    exec -ignorestderr make -C ${worksrcpath}/cups
```
Comment (reneeotten) [inline print/cups-ppdc/Portfile:69]: I am not 100% sure, but I don't think we typically use `exec` for this kind of things... most Portfiles use  `system -W ${build.dir} "make -C ....."`, does that not work here?

File: print/cups-ppdc/Portfile [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    foreach tool {ppdc ppdi ppdpo ppdhtml ppdmerge} {
+        exec make -C ${worksrcpath}/man ${tool}.1 2>@1
+    }
+    exec make -C ${worksrcpath}/man ppdcfile.5 2>@1
+}
+
+destroot {
+    xinstall -d ${destroot}${prefix}/share/cups/ppdc
+
+    xinstall -m 644 -W ${worksrcpath}/data \
```
Comment (reneeotten) [inline print/cups-ppdc/Portfile:82]: please use `0644` and `0755` here and below

File: print/cups-ppdc/Portfile [hunk 5]
```diff
@@ -0,0 +1,100 @@
+# -*- coding: utf-8; mode: tcl; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- vim:fenc=utf-8:ft=tcl:et:sw=4:ts=4:sts=4
+
+PortSystem      1.0
+PortGroup       legacysupport 1.1
```
Comment (reneeotten) [inline print/cups-ppdc/Portfile:4]: why do you need to use the legacysupport PG? Typically there would be a comment for what missing function is needed and a specification on which systems

**Review by reneeotten — COMMENTED** [review]
a few comments based on a cursory look - there is too much stuff going on with patches that I do not fully understand, so I am hoping others will take a look as well...

File: print/cups-ppdc/Portfile (hunk 3 above)
Comment (evanmiller) [reply print/cups-ppdc/Portfile:69]: I was just figuring out how to do multiple make commands, will switch to `system`.

**Comment by evanmiller** [comment]
Regarding the patches, these two were accepted upstream:

patch-cups-hash-sha224.diff: https://github.com/OpenPrinting/cups/pull/225
patch-cups-usersys-ssl.diff: https://github.com/OpenPrinting/cups/pull/223

**Comment by evanmiller** [comment]
Updated with @reneeotten's suggestions and added logic to compress the man pages.

**Comment by reneeotten** [comment]
> patch-cups-hash-sha224.diff: https://github.com/OpenPrinting/cups/pull/225
> 
> patch-cups-usersys-ssl.diff: https://github.com/OpenPrinting/cups/pull/223

then please add these links in the respective patchfiles

**Comment by reneeotten** [comment]
> added logic to compress the man pages.

I *think* MacPorts should compress man pages automatically, is that not working here?

**Comment by evanmiller** [comment]
> I _think_ MacPorts should compress man pages automatically, is that not working here?

You are correct. I've removed those lines entirely.

**Comment by evanmiller** [comment]
> > patch-cups-hash-sha224.diff: [OpenPrinting/cups#225](https://github.com/OpenPrinting/cups/pull/225)
> > patch-cups-usersys-ssl.diff: [OpenPrinting/cups#223](https://github.com/OpenPrinting/cups/pull/223)
> 
> then please add these links in the respective patchfiles

Done

**Comment by evanmiller** [comment]
@reneeotten One of the patch files isn't applying as cleanly as I'd like ("fuzz 1") - let me fool around with it this afternoon and then I will give the all-clear.

**Comment by evanmiller** [comment]
`patch-static-targets.diff` is cleaned up now. Ready for merge from my perspective.

**Comment by reneeotten** [comment]
> `patch-static-targets.diff` is cleaned up now. Ready for merge from my perspective.

thank you @evanmiller - in it goes!

**Comment by reneeotten** [comment]
@evanmiller this fails on the buildbots for 10.6 ([i386](https://build.macports.org/builders/ports-10.6_i386-watcher/builds/21333) and [x86_64](https://build.macports.org/builders/ports-10.6_x86_64-watcher/builds/21304)) and [10.7](https://build.macports.org/builders/ports-10.7_x86_64-watcher/builds/21340). I am clueless about the reason why and/or possible fix for it, but perhaps you want to take a look?

**Comment by evanmiller** [comment]
@reneeotten Will do, thanks for the heads-up. 10.6 looks like a code signing issue (easy fix), the 10.7 error I will need to research.

**Comment by evanmiller** [comment]
This is where the configure script is failing on 10.7:

https://github.com/apple/cups/blob/9d4e1ef79a96008a747ddaab9578157db4c65515/config-scripts/cups-common.m4#L395-L404

I haven't heard of this issue before – I guess `<sandbox/private.h>` is wrongly included in a public header on 10.7? I can try commenting out the check and hoping for the best. It's especially strange that the check requires this file in `/usr/local` specifically.

**Comment by reneeotten** [comment]
If it's too much trouble you don't have to get ports to build on old systems, so you could leave it broken. But then it would be good to error out early with a block like this (not adapted for correct versions):
```
if {${os.platform} eq "darwin" && ${os.major} < 16} {
    known_fail      yes
    pre-fetch {
        # last version that runs on Lion is 3.6.1
        # see https://trac.macports.org/ticket/58504
        ui_error "${subport} requires OS X 10.12 or later"
        return -code error "incompatible OS version"
    }
}
```

**Comment by evanmiller** [comment]
@reneeotten Thanks. The tools are more useful on older systems (I'm running 10.4) so I'd like to support them. I'll open a new PR with proposed fixes.

**Comment by reneeotten** [comment]
> @reneeotten Thanks. The tools are more useful on older systems (I'm running 10.4) so I'd like to support them. I'll open a new PR with proposed fixes.

I figured that you had some interest in older systems ;)

**Pull request merged**